### PR TITLE
trace: Runtime filter message handler

### PR DESCRIPTION
**Pull request by ktrzcinx**
It's a FW part extracted from https://github.com/thesofproject/sof/pull/2965. This part have ABI changes.

Kernel part: https://github.com/thesofproject/linux/pull/2377

### CODE REVIEW

File: src/ipc/handler.c [hunk 1]
```diff
@@ -747,6 +747,47 @@ static int ipc_dma_trace_config(uint32_t header)
...
+	struct sof_ipc_trace_filter_elem *end;
+	struct sof_ipc_trace_filter *packet;
+	struct trace_filter filter;
+	struct ipc *ipc = ipc_get();
+	int ret = 0;
+	int cnt;
+
+	packet = ipc->comp_data;
+	elem = packet->elems;
+	end = &packet->elems[packet->elem_cnt];
```
Comment (lgirdwood) [inline src/ipc/handler.c:763]: Can we validate end and elem_cnt is not overflowing msg size and mailbox.

File: src/trace/trace.c [hunk 2]
```diff
@@ -142,9 +142,100 @@ struct sof_ipc_trace_filter_elem *trace_filter_fill(struct sof_ipc_trace_filter_
...
+	return cnt;
+}
+
 int trace_filter_update(const struct trace_filter *filter)
 {
-	return -EINVAL;
+	int ret = 0;
+
+	/* validate log level */
+	if (filter->log_level < LOG_LEVEL_CRITICAL ||
```
Comment (lgirdwood) [inline src/trace/trace.c:226]: Not following the comment here, these look like valid log levels ?

**Review by lgirdwood — CHANGES_REQUESTED** [review]
Almost good to go.

**Comment by lgirdwood** [comment]
@ktrzcinx can you link kernel PR here too.

File: src/trace/trace.c (hunk 2 above)
Comment (ktrzcinx) [reply src/trace/trace.c:226]: Here I assert, given log_level is in valid range. LOG_LEVEL_CRITICAL has low value, LOG_LEVEL_VERBOSE high

File: src/ipc/handler.c (hunk 1 above)
Comment (ktrzcinx) [reply src/ipc/handler.c:763]: added

**Review by lgirdwood — APPROVED** [review]
Kernel now merged, just waiting on CI....

**Comment by lgirdwood** [comment]
@ktrzcinx can you ping me when kernel part is good and we can merge this one.

**Comment by ktrzcinx** [comment]
@lgirdwood, of course, I will do that

**Review by kv2019i — CHANGES_REQUESTED** [review]
With IPC ABI hat on, some minor concerns raised in kernel side review, let's keep open until those are solved: -> https://github.com/thesofproject/linux/pull/2377#discussion_r487933085

**Comment by plbossart** [comment]
@ktrzcinx can you update the firmware part to reflect the changes made on the kernel side. I think it's quite good now, and it'd be a nice addition to firmware 1.6 and kernel 5.10.

**Comment by ktrzcinx** [comment]
@plbossart Aligned with kernel code.

**Review by kv2019i — APPROVED** [review]
ABI is now ok for kernel side as well. We are continuing with some codestyle issues for the kernel PR, but ABI part is good.

@lgirdwood As this missed the 1.6 release branch, should this bump ABI to 3.18? If yes, @ktrzcinx could you update the PR with annotations following the recently merged https://github.com/thesofproject/sof-docs/pull/283 .. ?

File: src/include/user/abi_dbg.h [hunk 3]
```diff
@@ -19,7 +19,7 @@
 #define __USER_ABI_DBG_H__
 
 #define SOF_ABI_DBG_MAJOR 5
-#define SOF_ABI_DBG_MINOR 1
+#define SOF_ABI_DBG_MINOR 2
```
Comment (lgirdwood) [inline src/include/user/abi_dbg.h:22]: @ktrzcinx can you confirm this is a debug ABI update only. If so, we can put in in ABI3.17.

Comment (ktrzcinx) [reply src/include/user/abi_dbg.h:22]: This change bumps both ABI numbers

**Comment by ktrzcinx** [comment]
>@lgirdwood As this missed the 1.6 release branch, should this bump ABI to 3.18? If yes, @ktrzcinx could you update the PR with annotations following the recently merged thesofproject/sof-docs#283 .. ?

ABI bumped, added ABI version information for new structures .

**Comment by lgirdwood** [comment]
@ktrzcinx thank you, btw I've cleaned up the ABI MINOR PRs and I think we are good for this is 3.17 (which we can also close for v1.6). Sorry to ask you to change the ABI back to 3.17 but I think we can get in in v1.6. @kv2019i do you agree ?

**Comment by ktrzcinx** [comment]
Updated to 3.17 version

**Comment by lgirdwood** [comment]
@kv2019i ping any comment here to keep thi sin ABI 3.17 for v1.6 ?

**Comment by kv2019i** [comment]
@lgirdwood wrote:
> @kv2019i ping any comment here to keep thi sin ABI 3.17 for v1.6 ?
Aa, you missed my thumbs-up reaction to https://github.com/thesofproject/sof/pull/3329#issuecomment-694229555 . But yes, go ahead with 3.17 and v1.6!

**Pull request merged**